Comment 3-Dot Menu With Edit and Remove Support

dougkoellmer opened the pull request:
## PR Checklist

<!-- For the checkbox formatting to work properly, make sure there are no spaces on either side of the "x" -->

Please check all that apply to this PR using "x":

- [x] I have checked that this PR is not a duplicate of an existing PR (open, closed or merged)
- [x] I have checked that this PR does not introduce a breaking change
- [ ] This PR introduces breaking changes and I have provided a detailed explanation below

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting)
- [ ] Refactoring (no functional changes)
- [ ] Documentation changes
- [ ] Other - Please describe:

## Fixes

Issue Number: #91 Task: `Need comment edit/delete options`

## What is the current behavior?

Only three options available on a given comment: `Like`, `Dislike`, and `Block`

## What is the new behavior?

Adds support for `Edit` and `Remove` through a new 3-dot kebab menu like web. Also stubs out support for the still missing commenting features: `Love`, `Pin`, `Add as moderator`, `Mute`, and `Copy Link`.

## Other information

- This includes some minor refactoring and bug fixes to related commenting features. Deeper refactoring can be done at a later point, to bring all Comment Actions under one consistent umbrella. Right now there are different UI hooks and networking calls for various comment actions.
- Fixed a bug where comment list would not display after making first comment on channel or video. This was already on `master` I believe.
- Fixed a crash if you tried to expand comment area before comment list loaded on a video.
- General PR feedback. See in comment below: https://github.com/OdyseeTeam/odysee-android/pull/116#issuecomment-1035619144

kekkyojin commented on app/src/main/java/com/odysee/app/utils/Comments.java:
```diff
@@ -24,32 +26,48 @@
+        // TODO: Experimented with this call since the above call is deprecated. I was trying to track down a signing bug and don't know if this
+        // helped or hurt since there were too many other variables at play, and now I don't want to rock the boat unnecessarily.
+        // Delete when decided on.
+//            return (JSONObject) Lbry.authenticatedGenericApiCall("channel_sign", signingParams, commentBody.getString("auth_token"));
```
authenticatedGenericApiCall is the right method to be used

dougkoellmer replied:
OK I'll try it out, thanks.

kekkyojin commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.1.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.2-all.zip
```
I am almost 100% sure that this change should be reverted. Mobile team is using a build server and they don’t change the version of this as often as developers do in our local machines.

From what I have seen, there is no need to update Gradle or Android Gradle Plugin.

dougkoellmer replied:
Will do!

kekkyojin commented on app/src/main/java/com/odysee/app/tasks/CommentOptionTask.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Performs the network call for a {@link CommentOption}
+ */
+public class CommentOptionTask extends AsyncTask<Void, Void, Void> {
```
DON’T use AsyncTask as they are deprecated and using threads along Futures or Supplier will work without the need of using artificially added classes like AsyncTask

dougkoellmer replied:
Agreed! I was copying similar code I saw that uses AsyncTask, erring on the side of consistency with existing code in the project. Perhaps I will come up with a generic replacement and replace AsyncTask in a few other places when I get back to this PR.

dougkoellmer commented:
Thanks for feedback @kekkyojin. I will address your feedback and still clean up a few minor things on my personal TODO list, including a bit more documentation. I can issue a formal PR tomorrow unless the team wants me to go ahead with the other missing commenting features, in which case I can submit a larger PR at the end of the week. Let me know!

akinwale reviewed (changes requested):
Ran into a couple of issues while testing:

* Crashes with this error on the channel comments view:

```
2022-02-10 03:16:11.080 11534-11534/com.odysee.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.odysee.app, PID: 11534
    java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.odysee.app.model.Claim.getClaimId()' on a null object reference
        at com.odysee.app.adapter.CommentOption$4.isAvailable(CommentOption.java:126)
        at com.odysee.app.adapter.CommentOption.areAnyOptionsAvailable(CommentOption.java:251)
        at com.odysee.app.adapter.CommentListAdapter.onBindViewHolder(CommentListAdapter.java:352)
        at com.odysee.app.adapter.CommentListAdapter.onBindViewHolder(CommentListAdapter.java:43)
```

* If we're removing the only (single) comment on the content, there should be no preview comment.
* When the comment list is empty after the removal of an only (single) comment, it should display the "no comments" message.

Other notes
* Can the edit comment option make use of the existing comment box instead of creating a new popup entirely?
* Everything else looks good so far.

dougkoellmer commented:
> Can the edit comment option make use of the existing comment box instead of creating a new popup entirely?

The tricky part is if there are a lot of comments and you scroll away from the one that is being edited, you have to maintain state in the adapter so that the EditText can be reconstructed with the edited text and same cursor position. Some questions of what to do with keyboard as well. You could freeze scrolling but if it's a long comment then you have some UX issues there as well.

So probably not impossible but in the past I've always addressed this situation with some kind of modal like a dialog. The YouTube app for example uses a Toast/Drawer from the bottom right above the keyboard.

akinwale commented:
> > Can the edit comment option make use of the existing comment box instead of creating a new popup entirely?
> 
> The tricky part is if there are a lot of comments and you scroll away from the one that is being edited, you have to maintain state in the adapter so that the EditText can be reconstructed with the edited text and same cursor position. Some questions of what to do with keyboard as well. You could freeze scrolling but if it's a long comment then you have some UX issues there as well.
> 
> So probably not impossible but in the past I've always addressed this situation with some kind of modal like a dialog. The YouTube app for example uses a Toast/Drawer from the bottom right above the keyboard.

Ok, that makes sense. In this case, let's use a modal BottomSheetDialog to present the edit comment interface.

dougkoellmer commented:
PR feedback in! To confirm:
- Removed usage of AsyncTask. Replaced with plain old Java thread like I saw elsewhere in the project.
- Switched to `authenticatedGenericApiCall` for signing comment-related requests.
- Rolled back gradle wrapper version.
- Made sure that transitions from comments to no comments and vice versa properly updates the UI.
- Could not replicate crash found by Akinwale but added null checks that should fix it.
- Fixed crash already in `master` if you try to expand comments before they're loaded.
- Using `BottomSheetDialogFragment` to edit comment text instead of standard dialog.
- For `If we're removing the only (single) comment on the content, there should be no preview comment.` I'm not totally sure what the issue is. I may have fixed it accidentally before trying to replicate. :) Let me know if there is still an issue.

kekkyojin commented on app/src/main/res/layout-sw360dp/dialog_edit_comment.xml:
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
```
App will crash for devices with a display smaller width of 360 dp. Maybe this could go on the layout/ folder instead

dougkoellmer replied:
Hm not sure how that happened. Not intentional! Good catch. :)

kekkyojin commented on app/src/main/res/layout-sw360dp/dialog_edit_comment.xml:
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Suggestion: Could you refactor this as a ConstraintLayout?

dougkoellmer commented:
Changed to ConstraintLayout for `dialog_edit_comment.xml`, let me know if you have any other feedback.

kekkyojin commented:
Thanks, @dougkoellmer. I was looking for what parts of the code will this PR change in order to avoid creating new PR which could collide when merging. Then I saw the problem with the layout-sw360 and I also added another comment about moving from relativeLayout to ConstraintLayout. It was after I added the comment to the review list when I realized that using ConstraintLayout is more of a personal choice, although I am also trying to use ConstraintLayout for my new code wherever it is possible.

I tried to remove the comment before posting my review, but I was unable, so I "resolved" it.

Thanks anyway. ConstraintLayout will not hurt the app and could make easier to perform future layout changes thanks to the visual editor.

dougkoellmer commented:
Thanks for your comments and feedback @kekkyojin. Yes ConstraintLayout can make more sense compared to RelativeLayout. They both have their quirks. :) 

More important is consistency so if you're using ConstraintLayout for new code then it makes sense to change it in this PR.

kekkyojin commented:
Sorry for the new conflicting file. I merged too quickly. Luckily it will be very easy to fix

akinwale reviewed (changes requested):
I tested the edit and delete and that looks good.

However, you seem to have removed the comments preview, which expands into the commenting area. This should not be the case.
The comment preview should always be present on the file page, regardless of whether comments exist or not.
Clicking on the expand control in the comment preview area should open up the commenting area with the post comment form and the list of comments.

dougkoellmer commented:
Hi @akinwale I checked `master` and it seems to behave the same as my branch when there are no comments on a video. To be more specific about the cases in `master`:
- `No existing comments`: Presents the "Post Comment" form. I sometimes see the expand control for a brief moment but then it disappears and presents only the "Post Comment" form.
- `Existing comments`: I only see the expand control and a preview of the first comment. You have to click the expand control in order to see other comments or post a new comment.

Sorry let me know if I'm missing something! In the meantime I'll fix up the conflicts here quickly.

dougkoellmer commented:
I just updated my fork to fix conflicts with `master`. I checked the comment expansion behavior again and my branch seems to behave the same as `master` but let me know if I'm missing something. Thanks.

akinwale reviewed (approved):
Looks good to me.

dougkoellmer commented:
Thanks for merging. :)
